### DISCUSSION TITLE
Method to determine whether all realm instances are closed

### DISCUSSION THREAD

**Discussion by peppy** [realm/realm-dotnet#2657] [Q&A] [answered]
In our application, we give the user the ability to migrate their data folder from one location to another. Before doing so, we generally try and ensure all file handles are closed. Until now, we've been using a manual approach where we wrap each async database usage in our own class and wait for all usages to finish.

With our switch to realm, we're trying to be as close-to-the-wire as possible (ie. we want to use `Realm` directly rather than a `RealmUsage` wrapper), but this makes it hard for us to know when all usages are finished. I know realm has pretty competent locking/awareness internally of all contexts process wide, and was wondering if there's a way for us to check on this.

ie. we'd be looking for something like `Realm.AnyInstanceOpen(realmConfiguration)`.

If such a method doesn't exist, I have two ideas about what we could do instead, but not sure how reliable they are yet:
- Call a method like `Realm.Compact` until it succeeds, implying all usages have finished (looks to work)
- Use reflection to check the `_state.liveRealms` or [something similar](https://github.com/realm/realm-dotnet/blob/32f4ebcc88b3e80a3b254412665340cd9f3bd6b5/Realm/Realm/Realm.cs#L1433), though obviously we would hope to avoid this

**Comment by LaPeste** [comment, marked as answer]
Hi @peppy,

There is no recommended way to do what you asked. In our public API, we also don't currently expose a method to move the realm.

I'd say that, as long as a solution fits the performance needs, any solution is fine. So, the `Realm.Compact` solution that you suggested is fine as long as you are OK with waiting for the instance getting compacted.
You could also create an extension method on `Realm` to obtain an instance and at the same time to save somewhere all the refs to the created instances. At that point you'd just need to query [`Realm.IsClosed`](https://docs.mongodb.com/realm-sdks/dotnet/latest/reference/Realms.Realm.html#Realms_Realm_IsClosed) on each known instance.

I hope this answers what you were looking for.

**Reply by peppy** [reply]
Thanks for the confirmation.

We've been using the compact solution in production for a while now and it seems solid. Performance is luckily not a concern as this is a rare operation (where the user is relocating the whole application).
